Tool to create a screenshot automatically using a webhook fixture

punchagan opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->

This PR adds a script to the `tools` directory to automatically create or update the screenshot for a webhook integration from a test fixture.

timabbott commented on tools/generate-docs-integration-screenshot:
```diff
@@ -0,0 +1,125 @@
+from zerver.lib.webhooks.common import get_fixture_http_headers
+
+def create_integration_bot(integration_name: str) -> UserProfile:
+    owner = UserProfile.objects.get(id=10)
```
We should probably use `get_user()` so this doesn't break if we ever have the IDs change.

timabbott commented on tools/generate-docs-integration-screenshot:
```diff
@@ -0,0 +1,125 @@
+        if logo_path.endswith(".svg"):
+            logo_png_path = "/tmp/{}-logo.png".format(integration_name)
+            cairosvg.svg2png(url=logo_path, write_to=logo_png_path)
+            logo_path = logo_png_path
```
This is fine for now, but I think it'd make sense to plan to move this logic into e.g. `tools/setup/generate_zulip_bots_static_files.py` (Or maybe a renamed `tools/setup/optimize-svg`, and just commit the PNGs alongside the SVGs?).

timabbott replied:
(We could just do a TODO comment for this detail to keep track of it)

punchagan replied:
The PNGs are only used by developers generating the documentation. I'm not sure if we really need to have them in the repo.

timabbott commented on tools/generate-docs-integration-screenshot:
```diff
@@ -0,0 +1,125 @@
+
+        if logo_path.endswith(".svg"):
+            logo_png_path = "/tmp/{}-logo.png".format(integration_name)
+            cairosvg.svg2png(url=logo_path, write_to=logo_png_path)
```
(zulip-py3-venv) tabbott@coset:~/zulip$ ./tools/generate-docs-integration-screenshot zerver/webhooks/github/fixtures/pull_request__assigned.json 
Traceback (most recent call last):
  File "./tools/generate-docs-integration-screenshot", line 45, in create_integration_bot
    bot = UserProfile.objects.get(email=bot_email)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/db/models/query.py", line 408, in get
    self.model._meta.object_name
zerver.models.DoesNotExist: UserProfile matching query does not exist.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/urllib/request.py", line 1474, in open_local_file
    stats = os.stat(localfile)
FileNotFoundError: [Errno 2] No such file or directory: '/home/tabbott/zulip/http:/localhost:9991/static/images/integrations/logos/github.svg'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./tools/generate-docs-integration-screenshot", line 105, in <module>
    bot = create_integration_bot(integration.name)
  File "./tools/generate-docs-integration-screenshot", line 64, in create_integration_bot
    cairosvg.svg2png(url=logo_path, write_to=logo_png_path)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/cairosvg/__init__.py", line 70, in svg2png
    output_height=output_height)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/cairosvg/surface.py", line 144, in convert
    **kwargs)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/cairosvg/parser.py", line 402, in __init__
    parse_url(self.url), 'image/svg+xml')
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/cairosvg/parser.py", line 258, in fetch_url
    return read_url(url, self.url_fetcher, resource_type)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/cairosvg/url.py", line 159, in read_url
    return url_fetcher(url, resource_type)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/cairosvg/url.py", line 100, in fetch
This line seems to require the development server running; does it support either `file://` URLs or just passing a path instead of doing a `urlopen`?    One gets this exception is the server isn't running.

  ```
  return urlopen(Request(url, headers=HTTP_HEADERS)).read()
  File "/usr/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.6/urllib/request.py", line 526, in open
    response = self._open(req, data)
  File "/usr/lib/python3.6/urllib/request.py", line 544, in _open
    '_open', req)
  File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 1452, in file_open
    return self.open_local_file(req)
  File "/usr/lib/python3.6/urllib/request.py", line 1491, in open_local_file
    raise URLError(exp)
urllib.error.URLError: <urlopen error [Errno 2] No such file or directory: '/home/tabbott/zulip/http:/localhost:9991/static/images/integrations/logos/github.svg'>
```

punchagan replied:
Looks like 99242138a725477cd44c09b1069ce8b3d9c21aec seems to have broken this (which I hadn't pulled, while working locally). Sorry for the trouble!

timabbott commented on tools/generate-docs-integration-screenshot:
```diff
@@ -0,0 +1,125 @@
+
+    return bot
+
+def get_integration_fixture_names(path: str) -> Tuple[str, str]:
```
I might call this `split_fixture_path` and put it in `zerver/lib/integrations.py`.

timabbott commented on tools/message-screenshot.js:
```diff
@@ -0,0 +1,66 @@
+        clip.height += 10;
+        const imagePath = path.join(__dirname, '..', 'static', 'images', 'integrations', options.integration, '001.png');
+        await page.screenshot({ path: imagePath, clip: clip });
+        console.log(`Captured screenshot for ${options.integration}!`);
```
This should print the path, maybe with some color, at the end so it's easy to find :)

timabbott commented on tools/message-screenshot.js:
```diff
@@ -0,0 +1,66 @@
+            '--font-render-hinting=none',
+        ],
+        defaultViewport: null,
+        headless: true,
```
Maybe worth adding some comments on these parameters and where it's important that they're different from the normal puppeteer parameters we use?  

Also there's probably a refactor we could do to have most of this share code with `frontend_tests/puppeteer_tests/00-realm-creation.js` so the outer `try/catch/` block is shared, for example.

timabbott commented:
This is awesome @punchagan!  I posted a bunch of comments; I'm happy to merge this prototype after whatever subset of those you feel are convenient to fix and then fix-forward from there.

The ideal long-term version of this would be something where we replace every webhook integration screenshot in the codebase with one generated using this tool, which would (A) fix all of those that have old designs in them and (B) set us up to be able to regenerate the screenshots freely in the future as we adjust the visual design, in addition to streamlining the process for making new integrations.

But I think we should focus on making this base feel clean and nicely integrated into the codebase before working too much on doing this in a giant loop (And deciding how to declare which fixtures to use, etc.).

punchagan commented:
@timabbott I've re-pushed to this branch addressing most of your comments (except for the code-sharing with puppeteer tests, for which I added a TODO). 

I had similar ideas, when I started playing around with this code, to your long term vision. I agree that focusing on making this base cleaner and well integrated would be a good starting point.

timabbott commented:
I got this error the first time I ran it; it needed to be fixed by running `test-js-with-puppeteer`.  Something we'll naturally fix when we migrate to sharing code.

`(node:22428) UnhandledPromiseRejectionWarning: Error: Browser is not downloaded. Run "npm install" or "yarn install"`

timabbott commented:
This is a great checkpoint, merged, thanks @punchagan!  

I guess is the next step to update our docs for making integrations?

The other thing I notice is that this fails if `run-dev.py` isn't running; we should either improve the error message or making it use `test_server.py` and run its own.
